0.12 SDK / Test Runner can't use exclusively aliased provider

### Terraform Version

```
Vendored 247faca59d9d (0.12, master a few days ago) into the `google-beta` provider
```

### Terraform Configuration Files
https://github.com/terraform-providers/terraform-provider-google-beta/blob/master/google-beta/resource_compute_managed_ssl_certificate_generated_test.go

Also relevant:

https://github.com/terraform-providers/terraform-provider-google-beta/blob/master/google-beta/provider_test.go#L69-L76

### Expected Behavior
The test runs to completion, as it does in 0.11

### Actual Behavior
```
        Error: Error initializing context: 2 problems:
        
        - Could not satisfy plugin requirements: 
        Plugin reinitialization required. Please run "terraform init".
        
        Plugins are external binaries that Terraform uses to access and manipulate
        resources. The configuration provided requires plugins which can't be located,
        don't satisfy the version constraints, or are otherwise incompatible.
        
        Terraform automatically discovers provider requirements from your
        configuration, including providers used in child modules. To see the
        requirements and constraints from each module, run "terraform providers".
        
        - provider "google" is not available
```

### Steps to Reproduce
Running `TestAccComputeManagedSslCertificate_managedSslCertificateBasicExample`

### Additional Context
We use the aliased provider to make sure that the content of the test will work with live Terraform; this worked in `0.11`, but I guess something has changed in `0.12`.

We'd strongly prefer a solution that doesn't involve also adding the `google` provider for these tests; we want to ensure the alias is used for every resource, since inside the `google-beta` tests the `google` provider *is* `google-beta`, confusingly, and configs could be incorrect.

In my local testing it looks like the problem arises only at the final destroy step of the test:

```
--- FAIL: TestAccComputeManagedSslCertificate_managedSslCertificateBasicExample (26.74s)
    testing.go:629: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: Error initializing context: 2 problems:
        
        - Could not satisfy plugin requirements: 
        Plugin reinitialization required. Please run "terraform init".
        
        Plugins are external binaries that Terraform uses to access and manipulate
        resources. The configuration provided requires plugins which can't be located,
        don't satisfy the version constraints, or are otherwise incompatible.
        
        Terraform automatically discovers provider requirements from your
        configuration, including providers used in child modules. To see the
        requirements and constraints from each module, run "terraform providers".
        
        - provider "google" is not available
```

This suggests that it's something in the _state_ that is causing this errant extra dependency. I'm going to dig further to see if I can figure out exactly what about the state (or the logic for extracting dependencies from it) is incorrect here.

--- a/helper/resource/state_shim.go
+++ b/helper/resource/state_shim.go
@@ -62,6 +62,7 @@ func shimNewState(newState *states.State, providers map[string]terraform.Resourc
 						Attributes: flatmap,
 						Tainted:    i.Current.Status == states.ObjectTainted,
 					},
+					Provider: res.ProviderConfig.String(),
 				}
 				if i.Current.SchemaVersion != 0 {
 					resState.Primary.Meta = map[string]interface{}{

--- a/helper/resource/state_shim_test.go
+++ b/helper/resource/state_shim_test.go
@@ -192,7 +192,8 @@ func TestStateShim(t *testing.T) {
 				},
 				Resources: map[string]*terraform.ResourceState{
 					"test_thing.baz": &terraform.ResourceState{
-						Type: "test_thing",
+						Type:     "test_thing",
+						Provider: "provider.test",
 						Primary: &terraform.InstanceState{
 							ID: "baz",
 							Attributes: map[string]string{
@@ -202,7 +203,8 @@ func TestStateShim(t *testing.T) {
 						},
 					},
 					"test_thing.foo": &terraform.ResourceState{
-						Type: "test_thing",
+						Type:     "test_thing",
+						Provider: "provider.test",
 						Primary: &terraform.InstanceState{
 							ID: "foo",
 							Attributes: map[string]string{
@@ -221,7 +223,8 @@ func TestStateShim(t *testing.T) {
 				Path: []string{"root", "child"},
 				Resources: map[string]*terraform.ResourceState{
 					"test_thing.baz": &terraform.ResourceState{
-						Type: "test_thing",
+						Type:     "test_thing",
+						Provider: "module.child.provider.test",
 						Primary: &terraform.InstanceState{
 							ID: "bar",
 							Attributes: map[string]string{
@@ -241,7 +244,8 @@ func TestStateShim(t *testing.T) {
 						Dependencies: []string{"data.test_data_thing.foo"},
 					},
 					"data.test_data_thing.foo": &terraform.ResourceState{
-						Type: "test_data_thing",
+						Type:     "test_data_thing",
+						Provider: "module.child.provider.test",
 						Primary: &terraform.InstanceState{
 							ID: "bar",
 							Attributes: map[string]string{
@@ -251,7 +255,8 @@ func TestStateShim(t *testing.T) {
 						},
 					},
 					"test_thing.lots.0": &terraform.ResourceState{
-						Type: "test_thing",
+						Type:     "test_thing",
+						Provider: "module.child.provider.test",
 						Primary: &terraform.InstanceState{
 							ID: "0",
 							Attributes: map[string]string{
@@ -261,7 +266,8 @@ func TestStateShim(t *testing.T) {
 						},
 					},
 					"test_thing.lots.1": &terraform.ResourceState{
-						Type: "test_thing",
+						Type:     "test_thing",
+						Provider: "module.child.provider.test",
 						Primary: &terraform.InstanceState{
 							ID: "1",
 							Attributes: map[string]string{
@@ -272,7 +278,8 @@ func TestStateShim(t *testing.T) {
 						},
 					},
 					"test_thing.single_count": &terraform.ResourceState{
-						Type: "test_thing",
+						Type:     "test_thing",
+						Provider: "module.child.provider.test",
 						Primary: &terraform.InstanceState{
 							ID: "single",
 							Attributes: map[string]string{
@@ -314,6 +321,6 @@ func TestStateShim(t *testing.T) {
 	}
 
 	if !expected.Equal(shimmed) {
-		t.Fatalf("\nexpected state:\n%s\n\ngot state:\n%s", expected, shimmed)
+		t.Fatalf("wrong result state\ngot:\n%s\n\nwant:\n%s", expected, shimmed)
 	}
 }

